File: code/string-algorithms/lapindrome_checker/lapindrome_checker.cpp
// Part of cosmos from OpenGenus Foundation

// including header files and liberaries
#include <iostream>
#include <algorithm>
using namespace std;

/*
  Function that will check if string is lapindrome or not
*/
bool is_lapindrome(string str)
{
  // calculating length
  int len = str.length();

  // getting mid element
  int mid = len/2;

  // getting delimiter
  int delim = len%2==0 ? mid : mid + 1;  // for odd number of characters we have to ignore the middle most element

  // slicing
  string left = str.substr(0,mid);
  string right = str.substr(delim, len);

  // sorting two strings
  sort(left.begin(), left.end());
  sort(right.begin(), right.end());

  // equating and returning
  if(left==right)
    return true;
  else
    return false;
}

/*
  Main Function
*/
int main()
{
  // case 1
  if(is_lapindrome("aabaa"))
    cout<<"lapindrome"<<endl;
  else
    cout<<"not lapindrome"<<endl;

  // case 2
  if(is_lapindrome("aabab"))
    cout<<"lapindrome"<<endl;
  else
    cout<<"not lapindrome"<<endl;

    return 0;

}

File: code/string-algorithms/lapindrome_checker/lapindrome_checker.py
# -----------------------------
#  Part of cosmos from OpenGenus Foundation
# -----------------------------

# function to perform algo and check if string is lapindrome or not
def is_lapindrome(string):
    # getting string length
    length = len(string)

    # getting mid
    mid = length // 2   # integer division done in order to prevent floating number

    # getting delimiter
    delim = mid # for even
    if length % 2 == 1:
        delim +=1 # for odd -> in odd length we have to ignore the middle most value

    # slicing string
    left = string[:mid]
    right = string[delim:]

    # sorting  string
    left_sorted = sorted(left) # this will return a list
    right_sorted = sorted(right) # this will return a list

    # converting list to string
    left = "".join(left_sorted)
    right = "".join(right_sorted)

    # equating
    if left == right:
        return True
    else:
        return False



"""
DEMONSTRATING
"""
# CASE 1
if(is_lapindrome("abba")):
    print("lapindrome")
    pass
else:
    print("not lapindrome")
    pass

# CASE 2
if(is_lapindrome("aaabba")):
    print("lapindrome")
    pass
else:
    print("not lapindrome")
    pass
